### ISSUE TITLE
problems with new lineage data and its start and end times

### ISSUE THREAD

**Issue by boris-dimitrov** [chanzuckerberg/idseq-web#796]
1) Missing genus names.  Bug in the scripts?

2) The BETWEEN operator is not the correct operator to select rows whose start/end time contains the taxon_count row time, because "t BETWEEN start AND end" is equivalent to "start <= t <= end" and what we actually need here is "start <= t < end" (the interval must be half-open).

3) Renaming a genus is highly problematic.  The current implementation permits rows with the old genus_name and the new genus_name to both be in effect -- with overlapping start/end times.  This means at the time we run genus_aggregation there is no uniquely defined name for the genus.  In fact, this causes every job to fail after host alignment, because the query in genus_aggregation fails with duplicate key update because of this. 

This approach of storing incremental version history in the table is not really feasible unless we normalize the schema and split out the name table again.

My recommendation is to revert this change until it better thought through.

cc @yunfangjuan 

```
mysql> select distinct(genus_name), genus_taxid, started_at, ended_at from taxon_lineages where genus_taxid in (65013, 81736, 518515, 337498, -200)  order by genus_taxid;
+-------------+-------------+---------------------+---------------------+
| genus_name  | genus_taxid | started_at          | ended_at            |
+-------------+-------------+---------------------+---------------------+
| NULL        |        -200 | 2000-01-01 00:00:00 | 3000-01-01 00:00:00 |
|             |        -200 | 2000-01-01 00:00:00 | 2018-02-23 00:58:30 |
|             |        -200 | 2018-02-23 00:58:30 | 2018-02-23 01:17:29 |
|             |        -200 | 2018-02-23 01:17:29 | 3000-01-01 00:00:00 |
|             |       65013 | 2000-01-01 00:00:00 | 2018-02-23 00:58:30 |
|             |       65013 | 2018-02-23 01:17:29 | 3000-01-01 00:00:00 |
| Marquesia   |       65013 | 2018-02-23 00:58:30 | 2018-02-23 01:17:29 |
|             |       81736 | 2000-01-01 00:00:00 | 2018-02-23 00:58:30 |
|             |       81736 | 2018-02-23 01:17:29 | 3000-01-01 00:00:00 |
| Barnardia   |       81736 | 2018-02-23 00:58:30 | 2018-02-23 01:17:29 |
|             |      337498 | 2000-01-01 00:00:00 | 2018-02-23 00:58:30 |
|             |      337498 | 2018-02-23 01:17:29 | 3000-01-01 00:00:00 |
| Liphanthus  |      337498 | 2018-02-23 00:58:30 | 2018-02-23 01:17:29 |
|             |      518515 | 2000-01-01 00:00:00 | 2018-02-23 00:58:30 |
|             |      518515 | 2018-02-23 01:17:29 | 3000-01-01 00:00:00 |
| Parachorius |      518515 | 2018-02-23 00:58:30 | 2018-02-23 01:17:29 |
+-------------+-------------+---------------------+---------------------+
16 rows in set (1.20 sec)
```

**Comment by boris-dimitrov** [comment]
```
mysql> select distinct(genus_name), genus_taxid, ended_at from taxon_lineages where ended_at='3000-01-01 00:00:00' and genus_taxid=-200;
+------------+-------------+---------------------+
| genus_name | genus_taxid | ended_at            |
+------------+-------------+---------------------+
| NULL       |        -200 | 3000-01-01 00:00:00 |
|            |        -200 | 3000-01-01 00:00:00 |
+------------+-------------+---------------------+
2 rows in set (0.00 sec)

```

**Comment by boris-dimitrov** [comment]
cc @yunfangjuan

**Comment by boris-dimitrov** [comment]
Also this function doesn't appear to be updated to use the start/end times from the lineage table.
```
  def fetch_lineage_info(pipeline_run_id)
    lineage_records = TaxonLineage.where(
      "taxid in (select tax_id from taxon_counts
                 where pipeline_run_id = #{pipeline_run_id}
                   and tax_level = #{TaxonCount::TAX_LEVEL_SPECIES})"
    )
    result_map = {}
    search_key_list = Set.new
    sort_map = {}
    lineage_records.each do |lr|
      key_array = []
      TaxonCount::NAME_2_LEVEL.each do |category, level|
        tax_name = lr["#{category}_name"]
        tax_id = lr["#{category}_taxid"]
        next unless tax_name && tax_name.strip.present?
        display_name = "#{tax_name} (#{category})"
        sort_key = "#{(10 - level)}-#{tax_name}"
        search_id = level * TAXON_CATEGORY_OFFSET + tax_id
        sort_map[search_id] = sort_key
        key_array << search_id
        search_key_list.add([display_name, search_id])
      end
      result_map[lr.taxid] = key_array
    end
    search_key_list = search_key_list.sort_by { |u| sort_map[u[1]] }
    { lineage_map: result_map, search_list: search_key_list }
  end
```

**Comment by boris-dimitrov** [comment]
These defaults are a bad idea:
```
  `started_at` datetime NOT NULL DEFAULT '2000-01-01 00:00:00',
  `ended_at` datetime NOT NULL DEFAULT '3000-01-01 00:00:00',
```
It essentially guarantees duplicate rows with overlapping timestamps.
These columns should NOT have defaults.

**Comment by boris-dimitrov** [comment]
Currently in production everything is fine because
```
mysql> select count(distinct(genus_name)) as c, genus_taxid from taxon_lineages group by genus_taxid order by c desc limit 1; 
+---+-------------+
| c | genus_taxid |
+---+-------------+
| 1 |     1985329 |
+---+-------------+
1 row in set (3.97 sec)
```
but alpha is broken for the reasons mentioned above.

**Comment by boris-dimitrov** [comment]
```
mysql> select count(*), started_at from taxon_lineages where ended_at='3000-01-01 00:00:00' and genus_taxid=-200 and genus_name IS NOT NULL group by started_at limit 10;
+----------+---------------------+
| count(*) | started_at          |
+----------+---------------------+
|    12392 | 2018-02-23 01:17:29 |
+----------+---------------------+
1 row in set (0.96 sec)

mysql> select count(*), started_at from taxon_lineages where ended_at='3000-01-01 00:00:00' and genus_taxid=-200 and genus_name IS NULL group by started_at limit 10;
+----------+---------------------+
| count(*) | started_at          |
+----------+---------------------+
|   355331 | 2000-01-01 00:00:00 |
+----------+---------------------+
1 row in set (1.14 sec)
```

**Comment by boris-dimitrov** [comment]
The row count in the lineages table on production is 1651841.     On alpha, that's the count for rows that end in the year 3000.

```
mysql> select count(1) from taxon_lineages where started_at='2000-01-01 00:00:00' and ended_at='3000-01-01 00:00:00';
+----------+
| count(1) |
+----------+
|  1535447 |
+----------+
1 row in set (0.92 sec)

mysql> select count(1) from taxon_lineages where ended_at='3000-01-01 00:00:00';
+----------+
| count(1) |
+----------+
|  1651841 |
+----------+
1 row in set (0.47 sec)

```

**Comment by boris-dimitrov** [comment]
The following query shows that the only seriously affected genus_taxid is -200 and there are 367723 taxons in that genus.
```
mysql> select count(distinct(IF(genus_name IS NULL, "NULL", genus_name))) as c, count(distinct(taxid)) as t,  genus_taxid from taxon_lineages where ended_at='3000-01-01 00:00:00' group by genus_taxid order by c desc limit 10;
+---+--------+-------------+
| c | t      | genus_taxid |
+---+--------+-------------+
| 2 | 367723 |        -200 |
| 1 |      2 |      115672 |
| 1 |      2 |     1649458 |
| 1 |      2 |     2038489 |
| 1 |      2 |      202161 |
| 1 |     10 |       47144 |
| 1 |      2 |      880240 |
| 1 |      6 |     1369733 |
| 1 |      2 |      328843 |
| 1 |     22 |      511126 |
+---+--------+-------------+
10 rows in set (5.92 sec)
```
The following 3 queries show those 367723 taxons are partitioned into two disjoint groups of 12392 and 355331 taxons respectively by (started_at, genus_name).   
```
mysql> select IF(genus_name IS NULL, "NULL", genus_name) as c, count(distinct(taxid)), count(1) from taxon_lineages where ended_at='3000-01-01 00:00:00' and genus_taxid = -200;
+------+------------------------+----------+
| c    | count(distinct(taxid)) | count(1) |
+------+------------------------+----------+
| NULL |                 367723 |   367723 |
+------+------------------------+----------+
1 row in set (1.10 sec)

mysql> select IF(genus_name IS NULL, "NULL", genus_name) as c, count(distinct(taxid)), count(distinct(started_at)), started_at from taxon_lineages where ended_at='3000-01-01 00:00:00' and genus_taxid = -200 and genus_name IS NOT NULL;
+------+------------------------+-----------------------------+---------------------+
| c    | count(distinct(taxid)) | count(distinct(started_at)) | started_at          |
+------+------------------------+-----------------------------+---------------------+
|      |                  12392 |                           1 | 2018-02-23 01:17:29 |
+------+------------------------+-----------------------------+---------------------+
1 row in set (0.95 sec)

mysql> select IF(genus_name IS NULL, "NULL", genus_name) as c, count(distinct(taxid)), count(distinct(started_at)), started_at from taxon_lineages where ended_at='3000-01-01 00:00:00' and genus_taxid = -200 and genus_name IS NULL;
+------+------------------------+-----------------------------+---------------------+
| c    | count(distinct(taxid)) | count(distinct(started_at)) | started_at          |
+------+------------------------+-----------------------------+---------------------+
| NULL |                 355331 |                           1 | 2000-01-01 00:00:00 |
+------+------------------------+-----------------------------+---------------------+
```

**Comment by boris-dimitrov** [comment]
Therefore the simplest workaround is
```
update taxon_lineages set genus_taxid=NULL where started_at = '2018-02-23 01:17:29' and genus_taxid IS NOT NULL limit 15000
```
and it will have worked correctly if it updates precisely 12392 rows and after that update the following query returns only "1" in column c:
```
select count(distinct(IF(genus_name IS NULL, "NULL", genus_name))) as c, count(distinct(taxid)) as t,  genus_taxid from taxon_lineages where ended_at='3000-01-01 00:00:00' group by genus_taxid order by c desc limit 10;
```

**Comment by cdebourcy** [comment]
- The taxon_lineages table is different on alpha and prod because I have only run the rake task update_lineage_db on alpha. This is because I wanted to hold off from doing it on prod until it is tested more thoroughly on alpha.
- Regarding the BETWEEN operator (point 2): I agree the interval should actually be half-open. However, this only makes a difference in the case where a taxon_count is created at exactly the same time as a record in the lineage table. This case does not arise because we currently wait until quiescence before pushing an index/lineage update.
- Regarding defaults for started_at and ended_at: we need these because we do in fact need to produce violations of unique indexes in order to overwrite existing records using mysqlimport when they need to be updated (see code of the rake task update_lineage_db). Conceptually, the defaults correspond to -infinity and +infinity. I believe this part was pretty well thought-out in the discussion between @yunfangjuan, @boris-dimitrov and myself, but let's revisit today.
- I believe the only actual bug here is with "missing" (negative) taxids like -200 -- thanks for pointing that out. Clearly this special case needs more consideration. Let's discuss today!

**Comment by cdebourcy** [comment]
Some context to help interpret the SQL output in the original post: I had run `rake update_lineage_db` several times on alpha, making small test changes, and then in the end I attempted to reset taxon_lineages to its original functionality by running `rake update_lineage_db` with the same reference files it had used originally when the table was first created. That's why you see 3 entries for taxid 65013, for example: 
- no name from the beginning until 2018-02-23 00:58:30
- name `Marquesia` from 2018-02-23 00:58:30 until 2018-02-23 01:17:29
- back to no name from 2018-02-23 01:17:29 until the end of the millennium.

The one bug was caused by the use of NULL versus "" in the original database (see taxid -200). Will add a first step in the rake task that converts all blanks to "".
